signingscript/test/test_task.py

```diff
@@ -1,64 +1,184 @@
+), (
+    ['invalid'], []
+)))
+def test_get_suitable_signing_servers(context, formats, expected):
+    expected_servers = []
+    for info in expected:
+        expected_servers.append(
+            SigningServer(*info)
+        )
+
+    assert stask.get_suitable_signing_servers(
+        context.signing_servers, TEST_CERT_TYPE,
+        formats
+    ) == expected_servers
 
 
-def test_no_error_is_reported_when_no_missing_url(context):
-    context.task = task_generator.generate_object()
-    validate_task_schema(context)
+# get_token {{{1
+@pytest.mark.asyncio
+@pytest.mark.parametrize('exc,contents', ((
+    ScriptWorkerTaskException, 'token'
+), (
+    None, ''
+), (
+    None, 'token'
+)))
+async def test_get_token(event_loop, mocker, tmpdir, exc, contents, context):
+
+    async def test_token(*args, **kwargs):
+        if exc:
+            raise exc("Expected exception")
+        return contents
+
+    output_file = os.path.join(tmpdir, "foo")
+    mocker.patch.object(aiohttp, "BasicAuth", new=noop_sync)
+    mocker.patch.object(stask, "retry_request", new=test_token)
+    if exc or not contents:
+        with pytest.raises(SigningServerError):
+            await stask.get_token(context, output_file, TEST_CERT_TYPE, ["gpg"])
+    else:
+        await stask.get_token(context, output_file, TEST_CERT_TYPE, ["gpg"])
+        with open(output_file, "r") as fh:
+            assert fh.read().rstrip() == contents
+
+
+# sign_file {{{1
+@pytest.mark.asyncio
+@pytest.mark.parametrize('signtool,format', ((
+    'signtool', 'gpg'
+), (
+    ['signtool'], 'gpg'
+)))
+async def test_sign_file(context, mocker, format, signtool, event_loop):
+    work_dir = context.config['work_dir']
+    path = os.path.join(work_dir, 'filename')
+
+    async def test_cmdln(command):
+        assert command == [
+            'signtool', "-v",
+            "-n", os.path.join(work_dir, "nonce"),
+            "-t", os.path.join(work_dir, "token"),
+            "-c", context.config['ssl_cert'],
+            "-H", "127.0.0.1:9110",
+            "-f", format,
+            "-o", path, path,
+        ]
 
+    context.config['signtool'] = signtool
+    mocker.patch.object(stask, '_execute_post_signing_steps', new=noop_async)
+    mocker.patch.object(stask, '_execute_subprocess', new=test_cmdln)
+    await stask.sign_file(context, path, TEST_CERT_TYPE, [format], context.config['ssl_cert'])
 
+
+# _execute_post_signing_steps {{{1
 @pytest.mark.asyncio
-async def test_execute_post_signing_steps(context, monkeypatch):
+@pytest.mark.parametrize('suffix', ('apk', 'zip'))
+async def test_execute_post_signing_steps(context, monkeypatch, suffix):
     work_dir = context.config['work_dir']
-    abs_to = os.path.join(work_dir, 'target.apk')
+    abs_to = os.path.join(work_dir, 'target.{}'.format(suffix))
 
     async def zip_align_apk_mock(_context, _abs_to):
+        if suffix != 'apk':
+            assert 1 == 0  # We shouldn't call this on non-apk
```

Comment:
Nit: `assert 1 == 0` could be translated into something more self-explanatory, like `assert False` or `raise AssertionError`)